cpp/src/replace/replace.cu

```diff
@@ -138,64 +124,31 @@ __device__ T get_new_value(gdf_size_type         idx,
 
       if (input_is_valid) {
-        temp_data[threadIdx.x] = get_new_value<T, replacement_has_nulls>(i, input_data,
+        thrust::tie(output_data[i], output_is_valid)  =
+            get_new_value<T, replacement_has_nulls>(i, input_data,
                                       values_to_replace_begin,
                                       values_to_replace_end,
                                       d_replacement_values,
-                                      replacement_valid,
-                                      output_is_valid);
+                                      replacement_valid);
       }
 
     } else {
-       temp_data[threadIdx.x] = get_new_value<T, replacement_has_nulls>(i, input_data,
+       thrust::tie(output_data[i], output_is_valid) =
+            get_new_value<T, replacement_has_nulls>(i, input_data,
                                       values_to_replace_begin,
                                       values_to_replace_end,
                                       d_replacement_values,
-                                      replacement_valid,
-                                      output_is_valid);
+                                      replacement_valid);
     }
 
-    __syncthreads(); // waiting for temp_data to be ready
-
-    output_data[i] = temp_data[threadIdx.x];
-
-    /* output null counts calculations*/
+    /* output valid counts calculations*/
     if (input_has_nulls or replacement_has_nulls){
 
       bitmask &= __ballot_sync(active_mask, output_is_valid);
 
-      // allocating the shared memory for calculating the block_sum and
-      // block_valid_counts
-      __shared__ uint32_t warp_smem[warp_size];
-
-      if(threadIdx.x < warp_size) warp_smem[threadIdx.x] = 0;
-      __syncthreads();
-
-      const int wid = threadIdx.x / warp_size;
-      const int lane = threadIdx.x % warp_size;
-
-      if(lane == 0)warp_smem[wid] = __popc(active_mask);
-      __syncthreads(); // waiting for the sum of each warp to be ready
-
-      uint32_t block_sum = sum_warps<uint32_t>(warp_smem);
-
-      // reusing the same shared memory for block valid counts
-      if(threadIdx.x < warp_size) warp_smem[threadIdx.x] = 0;
-      __syncthreads();
-
-      if(lane == 0 && bitmask != 0){
-        int valid_index = wid + (blockIdx.x * (BLOCK_SIZE/warp_size));
-        output_valid[valid_index] = bitmask;
-        warp_smem[wid] = __popc(bitmask);
-      }
-      __syncthreads(); // waiting for the valid counts of each warp to be ready
-
-      // Compute total valid count for this block and add it to global count
-      uint32_t block_valid_count = sum_warps<uint32_t>(warp_smem);
-
-      // one thread computes and adds to null count
-      if (threadIdx.x < warp_size && lane == 0) {
-          atomicAdd(output_null_count, block_sum - block_valid_count);
+      if(threadIdx.x % warp_size == 0 && bitmask != 0){
+        output_valid[(int)(i/warp_size)] = bitmask;
+        atomicAdd(output_valid_count, __popc(bitmask));
```

Comment:
Hi @AK-ayush, I think you misunderstood.  We weren't suggesting you get rid of the shared memory reduction for valid_count, just the one for the block_count. Doing the atomicAdd for every warp will be slower than doing a single-warp reduction and a single atomicAdd.  So keep the previous shared memory + warp reduction for calculating valid count, but use the new `output.null_count = output.size - valids;` on the host side to compute total null_count.

Reply:
My apologies! I am updating right now...